File: problems/0105-construct-binary-tree-from-preorder-and-inorder-traversal/python/solution.py
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def to_level(root):
    if not root:
        return []
    out = []
    queue = deque([root])
    while queue:
        node = queue.popleft()
        if node:
            out.append(node.val)
            queue.append(node.left)
            queue.append(node.right)
        else:
            out.append(None)
    while out and out[-1] is None:
        out.pop()
    return out


def buildTree(preorder, inorder):
    """Construct Binary Tree from Preorder and Inorder — https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/"""
    index = {v: i for i, v in enumerate(inorder)}
    pre = [0]

    def helper(left, right):
        if left > right:
            return None
        val = preorder[pre[0]]
        pre[0] += 1
        node = TreeNode(val)
        mid = index[val]
        node.left = helper(left, mid - 1)
        node.right = helper(mid + 1, right)
        return node

    return helper(0, len(inorder) - 1)


def buildTreeArr(preorder, inorder):
    return to_level(buildTree(preorder, inorder))

File: problems/0105-construct-binary-tree-from-preorder-and-inorder-traversal/go/solution.go
package solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func toLevel(root *TreeNode) []*int {
	out := []*int{}
	if root == nil {
		return out
	}
	queue := []*TreeNode{root}
	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]
		if node != nil {
			v := node.Val
			out = append(out, &v)
			queue = append(queue, node.Left, node.Right)
		} else {
			out = append(out, nil)
		}
	}
	for len(out) > 0 && out[len(out)-1] == nil {
		out = out[:len(out)-1]
	}
	return out
}

// Construct Binary Tree from Preorder and Inorder — https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/
func buildTree(preorder, inorder []int) *TreeNode {
	index := make(map[int]int, len(inorder))
	for i, v := range inorder {
		index[v] = i
	}
	pre := 0
	var helper func(left, right int) *TreeNode
	helper = func(left, right int) *TreeNode {
		if left > right {
			return nil
		}
		val := preorder[pre]
		pre++
		node := &TreeNode{Val: val}
		mid := index[val]
		node.Left = helper(left, mid-1)
		node.Right = helper(mid+1, right)
		return node
	}
	return helper(0, len(inorder)-1)
}

func buildTreeArr(preorder, inorder []int) []*int {
	return toLevel(buildTree(preorder, inorder))
}
